pkg/agent/nodeportlocal/rules/iptable_rule.go

```diff
@@ -107,34 +108,29 @@ func (ipt *iptablesRules) AddAllRules(nplList []PodNodePort) error {
 }
 
 // DeleteRule deletes a specific NPL rule from NodePortLocalChain chain
-func (ipt *iptablesRules) DeleteRule(port int, podIP string) error {
-	klog.Infof("Deleting DNAT rule for Pod IP %s, port %d", podIP, port)
-	rule := buildRuleForPod(port, podIP)
-	if err := ipt.table.DeleteRule(iptables.NATTable, NodePortLocalChain, rule); err != nil {
+func (ipt *iptablesRules) DeleteRule(nodePort int, podIP string, podPort int) error {
+	podAddr := fmt.Sprintf("%s:%d", podIP, podPort)
+	rule := buildRuleForPod(nodePort, podAddr)
+	if err := ipt.table.DeleteRule(iptables.ProtocolIPv4, iptables.NATTable, NodePortLocalChain, rule); err != nil {
 		return err
 	}
+	klog.InfoS("Successfully deleted DNAT rule", "podAddr", podAddr, "nodePort", nodePort)
 	return nil
 }
 
 // DeleteAllRules deletes all NPL rules programmed in the node
 func (ipt *iptablesRules) DeleteAllRules() error {
-	exists, err := ipt.table.ChainExists(iptables.NATTable, NodePortLocalChain)
-	if err != nil {
-		return fmt.Errorf("failed to check if NodePortLocal chain exists in NAT table: %v", err)
-	}
-	if !exists {
-		return nil
+	if err := ipt.table.DeleteChain(iptables.ProtocolIPv4, iptables.NATTable, NodePortLocalChain); err != nil {
+		return err
```

Comment:
I did a test, deleting a chain which was still referenced by another rule always fails with "Too many links" error, so this would introduce a regression. Is this method covered by any test?

Reply:
Did you do it manually? "Too many links" is expected in this way.

Reply:
Anyway, it seems `DeleteChain` is only used in `DeleteAllRules` and `DeleteAllRules` is used by no one. I am going to delete them.

Reply:
I don't think it's expected. The original code deletes the chain at last so it ensures the chain is not referenced by any rule when deleting it. This patch changes to delete the chain in the beginning so it would introduce a regression. I'm not why there there is an orphaned method here, maybe it's left for historical reason, or maybe it's planed for future usage. Anyway it's beyond the scope of this PR. I would sugges to keep it as is if you want to merge this earlier. Otherwise I would invite the feature owner to comment whether deleting it is expected. If you keep the method, I would expect it wouldn't introduce a bug by this PR.

Reply:
`DeleteAllRules` and `ChainExists` are reverted.